Sources/SwiftCompilerPlugin/CompilerPlugin.swift

```diff
@@ -0,0 +1,246 @@
+    setvbuf(stdout, nil, _IONBF, 0)
+    #else
+    setvbuf(stdout, nil, _IOLBF, 0)
+    #endif
+
+    // Open a message channel for communicating with the plugin host.
+    pluginHostConnection = PluginHostConnection(
+      inputStream: FileHandle(fileDescriptor: inputFD),
+      outputStream: FileHandle(fileDescriptor: outputFD)
+    )
+
+    // Handle messages from the host until the input stream is closed,
+    // indicating that we're done.
+    let instance = Self()
+    do {
+      while let message = try pluginHostConnection.waitForNextMessage() {
+        try instance.handleMessage(message)
+      }
+    } catch {
+      // Emit a diagnostic and indicate failure to the plugin host,
+      // and exit with an error code.
+      internalError(String(describing: error))
+    }
+  }
+
+  // Private function to report internal errors and then exit.
+  fileprivate static func internalError(_ message: String) -> Never {
+    fputs("Internal Error: \(message)\n", stderr)
+    exit(1)
+  }
+
+  // Private function to construct an error message from an `errno` code.
+  fileprivate static func describe(errno: Int32) -> String {
+    if let cStr = strerror(errno) { return String(cString: cStr) }
+    return String(describing: errno)
+  }
+
+  /// Handles a single message received from the plugin host.
+  fileprivate func handleMessage(_ message: HostToPluginMessage) throws {
+    switch message {
+    case .getCapability:
+      try pluginHostConnection.sendMessage(
+        .getCapabilityResult(capability: PluginMessage.capability)
+      )
+      break
+
+    case .expandFreestandingMacro(let macro, let discriminator, let expandingSyntax):
+      try expandFreestandingMacro(
+        macro: macro,
+        discriminator: discriminator,
+        expandingSyntax: expandingSyntax
+      )
+
+    case .expandAttachedMacro(let macro, let macroRole, let discriminator, let customAttributeSyntax, let declSyntax, let parentDeclSyntax):
+      try expandAttachedMacro(
+        macro: macro,
+        macroRole: macroRole,
+        discriminator: discriminator,
+        customAttributeSyntax: customAttributeSyntax,
+        declSyntax: declSyntax,
+        parentDeclSyntax: parentDeclSyntax
+      )
+    }
+  }
+}
+
+/// Message channel for bidirectional communication with the plugin host.
+internal fileprivate(set) var pluginHostConnection: PluginHostConnection!
+
+typealias PluginHostConnection = MessageConnection<PluginToHostMessage, HostToPluginMessage>
+
+internal struct MessageConnection<TX, RX> where TX: Encodable, RX: Decodable {
+  let inputStream: FileHandle
+  let outputStream: FileHandle
+
+  func sendMessage(_ message: TX) throws {
+    // Encode the message as JSON.
+    let payload = try JSONEncoder().encode(message)
+
+    // Write the header (a 64-bit length field in little endian byte order).
+    var count = UInt64(payload.count).littleEndian
+    let header = Swift.withUnsafeBytes(of: &count) { Data($0) }
+    assert(header.count == 8)
```

Comment:
Just a few thought that crossed my mind. I’m fine with the way it is if you think it’s the best solution.

SourceKit-LSP uses the following HTTP-style header, which could allow us to add more options in the future (e.g. encoding of the message). I can’t really think of any options we want to add here but having the flexibility could come in handy in the future.

```
Content-Length: <some number in ASCII>\r\n\r\n
```

---

Alternatively, everything else in the communication channel is human readable, so I’m wondering whether it makes sense to also make this header human-readable. That might make debugging the plugin easier. I’m not sure if it’s worth it though.

Reply:
This basic message format is from SwiftPM's plugin mechanism (https://github.com/apple/swift-package-manager/blob/48192bbdd2800a38a82fa658a244e70b3167ec5d/Sources/PackagePlugin/Plugin.swift#L30-L31). Although the compiler plugin mechanism doesn't need to follow it, I thought it's good to have the same messaging format among the toolchain.

Reply:
Sure, makes sense to me. I just wanted to hear your thoughts on my ideas.